kuberay int with MCAD (#598)
ray-project/kuberay: docs/deploy/kuberay-with-MCAD.md

### Kuberay integration with MCAD (Multi-Cluster-App-Dispatcher)

> 

Author: Done, thanks

> The multi-cluster-app-dispatcher is a Kubernetes controller providing mechanisms for applications to manage batch jobs in a single or mult-cluster environment. For more details please refer [here](https://github.com/IBM/multi-cluster-app-dispatcher).

Author: Done, thanks

> In order to queue kuberay cluster and `gang dispatch` them when aggregated resources are available please refer to the setup [Kuberay-MCAD integration](https://github.com/IBM/multi-cluster-app-dispatcher/blob/quota-management/doc/usage/examples/kuberay/kuberay-mcad.md) with configuration files [here](https://github.com/IBM/multi-cluster-app-dispatcher/tree/quota-management/doc/usage/examples/kuberay/config).

Author: Done, thanks

# Kuberay integration with MCAD (Multi-Cluster-App-Dispatcher)

The multi-cluster-app-dispatcher is a Kubernetes controller providing mechanisms for applications to manage batch jobs in a single or multi-cluster environment. For more details please refer [here](https://github.com/IBM/multi-cluster-app-dispatcher).

## Use case

> MCAD allows you to deploy Ray cluster with a guarantee that sufficient resources are available in the cluster prior to actual pod creation in the kubernetes cluster. It supports features such as:

Author: Done, thanks

   
- Integrates with upstream Kubernetes scheduling stack for features such co-scheduling, Packing on GPU dimension etc.
- Ability to wrap any Kubernetes objects.

> - Increases control plane stability by JIT (Just-in Time) object  creation.

Reviewer: extra space
```suggestion
- Increases control plane stability by JIT (Just-in Time) object creation.
```

- Queuing with policies.
- Quota management that goes across namespaces.

> - Dispatching jobs to any one of the clusters.

Author: Yup, added Kubernetes.

> - Dispatching jobs to any one of the Kubernetes clusters.

Reviewer: Grammatically, it's slightly unclear what "the Kubernetes clusters" are.

For better context, perhaps it can be reworded like this:
`Support for multiple Kubernetes clusters; dispatching jobs to any one of a number of Kubernetes clusters.`

Author: Thanks, I will steal the edit and paste it to the docs :)

    Last Transition Micro Time:  2022-09-27T21:07:34.261174Z
    Last Update Micro Time:      2022-09-27T21:07:34.261174Z
    Reason:                      FrontOfQueue.
    Status:                      True
    Type:                        HeadOfLine
    Last Transition Micro Time:  2022-09-27T21:07:34.316208Z
    Last Update Micro Time:      2022-09-27T21:07:34.316208Z
    Reason:                      AppWrapperRunnable
    Status:                      True
    Type:                        Dispatched
  Controllerfirsttimestamp:      2022-09-27T21:07:34.251877Z
  Filterignore:                  true
  Queuejobstate:                 Dispatched
  Sender:                        before manageQueueJob - afterEtcdDispatching
  State:                         Running
Events:                          <none>
(base) asmalvan@mcad-dev:~/mcad-kuberay$ kubectl get pods
NAME                                               READY   STATUS    RESTARTS   AGE
raycluster-autoscaler-1-head-9s4x5                 2/2     Running   0          47s
raycluster-autoscaler-1-worker-small-group-4s6jv   1/1     Running   0          47s
```
- As seen the cluster is dispatched and pods are running.

> - Let's submit another kuberay cluster and see it queued without creating pending pods using command `kubectl create -f aw-raycluster.yaml`, note please change cluster name from `name: raycluster-autoscaler` to `name: raycluster-autoscaler-1` and re-submit

Reviewer: kuberay cluster -> Ray cluster

Author: Done, thanks

> - Let's submit another Ray cluster and see it queued without creating pending pods using command `kubectl create -f aw-raycluster.yaml`, note please change cluster name from `name: raycluster-autoscaler` to `name: raycluster-autoscaler-1` and re-submit

Reviewer: ```suggestion
- Let's submit another Ray cluster and see it queued without creating pending pods using the command `kubectl create -f aw-raycluster.yaml`. To do this, change the cluster name from `name: raycluster-autoscaler` to `name: raycluster-autoscaler-1` and re-submit.
```

Author: Done, thanks

    Type:                        Init
    Last Transition Micro Time:  2022-09-27T21:11:06.162401Z
    Last Update Micro Time:      2022-09-27T21:11:06.162401Z
    Reason:                      AwaitingHeadOfLine
    Status:                      True
    Type:                        Queueing
    Last Transition Micro Time:  2022-09-27T21:11:06.171619Z
    Last Update Micro Time:      2022-09-27T21:11:06.171618Z
    Reason:                      FrontOfQueue.
    Status:                      True
    Type:                        HeadOfLine
    Last Transition Micro Time:  2022-09-27T21:11:06.179694Z
    Last Update Micro Time:      2022-09-27T21:11:06.179689Z
    Message:                     Insufficient resources to dispatch AppWrapper.
    Reason:                      AppWrapperNotRunnable.
    Status:                      True
    Type:                        Backoff
  Controllerfirsttimestamp:      2022-09-27T21:11:06.161797Z
  Filterignore:                  true
  Queuejobstate:                 HeadOfLine
  Sender:                        before ScheduleNext - setHOL
  State:                         Pending
Events:                          <none>
```

> - As seen the second ray cluster is queued with no pending pods created. 

Reviewer: Ray should be capitalized

Author: Done, thanks

> - Dispatching policy out of the box is FIFO which can be augmented as per user needs. The second cluster will be dispatched when additional aggregated resources are available in the cluster or the first appwrapper kuberay cluster is deleted.

Reviewer: kuberay cluster -> Ray cluster

Does `appwrapper` have a preferred capitalization?

Author: Done, thanks

> - Dispatching policy out of the box is FIFO which can be augmented as per user needs. The second cluster will be dispatched when additional aggregated resources are available in the cluster or the first appwrapper kuberay cluster is deleted.

Reviewer: Is there a link you could provide here to options for MCAD scheduling policies?

Reviewer: ok, got it
